pkg/server/image_pull.go

```diff
@@ -26,195 +26,223 @@ import (
-		return desc, size, fmt.Errorf("failed to put %q: desc: %v err: %v", resolvedImageName, desc, err)
+	// Put the image information into containerd image store.
+	if err = c.imageStoreService.Put(ctx, ref, desc); err != nil {
+		return "", "", fmt.Errorf("failed to put image %q desc %v into containerd image store: %v",
+			ref, desc, err)
 	}
+	// TODO(random-liu): What if following operations fail? Do we need to do cleanup?
 
+	// Fetch all image resources into content store.
+	// Dispatch a handler for a sequence of handlers which:
+	// 1) fetch the object using a FetchHandler;
+	// 2) recurse through any sub-layers via a ChildrenHandler.
 	err = containerdimages.Dispatch(
 		ctx,
 		containerdimages.Handlers(
 			remotes.FetchHandler(c.contentIngester, fetcher),
 			containerdimages.ChildrenHandler(c.contentProvider)),
 		desc)
 	if err != nil {
-		return desc, size, fmt.Errorf("failed to fetch %q: desc: %v err: %v", resolvedImageName, desc, err)
+		return "", "", fmt.Errorf("failed to fetch image %q desc %+v: %v", ref, desc, err)
 	}
 
-	image, err := c.imageStoreService.Get(ctx, resolvedImageName)
+	// Read the image manifest from content store.
+	image, err := c.imageStoreService.Get(ctx, ref)
 	if err != nil {
-		return desc, size,
-			fmt.Errorf("get failed for image:%q err: %v", resolvedImageName, err)
+		return "", "", fmt.Errorf("failed to get image %q from containerd image store: %v", ref, err)
 	}
-	p, err := content.ReadBlob(ctx, c.contentProvider, image.Target.Digest)
+	digest := image.Target.Digest
+	p, err := content.ReadBlob(ctx, c.contentProvider, digest)
 	if err != nil {
-		return desc, size,
-			fmt.Errorf("readblob failed for digest:%q err: %v", image.Target.Digest, err)
+		return "", "", fmt.Errorf("readblob failed for digest %q: %v", digest, err)
 	}
 	var manifest imagespec.Manifest
-	err = json.Unmarshal(p, &manifest)
-	if err != nil {
-		return desc, size,
-			fmt.Errorf("unmarshal blob to manifest failed for digest:%q %v", image.Target.Digest, err)
+	if err := json.Unmarshal(p, &manifest); err != nil {
+		return "", "", fmt.Errorf("unmarshal blob to manifest failed for digest %q: %v", digest, err)
 	}
-	_, err = c.rootfsUnpacker.Unpack(ctx, manifest.Layers) // ignoring returned chainID for now
-	if err != nil {
-		return desc, size,
-			fmt.Errorf("unpack failed for manifest layers:%v %v", manifest.Layers, err)
+
+	// Unpack the image layers into snapshots.
+	if _, err = c.rootfsUnpacker.Unpack(ctx, manifest.Layers); err != nil {
+		return "", "", fmt.Errorf("unpack failed for manifest layers %v: %v", manifest.Layers, err)
 	}
-	size, err = image.Size(ctx, c.contentProvider)
+	// TODO(random-liu): Considering how to deal with the disk usage of content.
+
+	configDesc, err := image.Config(ctx, c.contentProvider)
 	if err != nil {
-		return desc, size,
-			fmt.Errorf("size failed for image:%q %v", image.Target.Digest, err)
+		return "", "", fmt.Errorf("failed to get config descriptor for image %q: %v", ref, err)
```

Comment:
Because we need `configDesc` later, so we could not assign it in with if clause.

If we don't care about the return value later, we could put them in one line.